oeps/oep-0015.rst

```diff
@@ -0,0 +1,144 @@
+
+(As an aside, it might be valuable for the existing script-tag feature to carry a similar warning, though that is outside the scope of this OEP.)
+
+Security was briefly touched on in the `initial edx-code thread`_ for this feature. One idea mentioned there (but not yet adopted in this OEP) is to allow only :code:`/static/*.js` scripts; this would prevent exploits in which an attacker compromises a server hosting an included script and replaces it with a malicious script, but it would also differ from the behavior of :code:`<script type="text/javascript">` tags which may use any valid :code:`https://` URL. Another comment raised the idea of a feature flag:
+
+    "We might want to make this feature be something that can be disabled if a given installation is not comfortable with giving this power to its authors. As you point out, the power is already there through multiple other mechanisms, so maybe this isn't a concern."
+
+.. _initial edx-code thread: https://groups.google.com/forum/#!topic/edx-code/T83TDxhH74E/discussion
+
+What about performance?
+-----------------------
+
+Several performance-related points were also raised via edx-code. The first concerns the use of this feature to load JavaScript libraries (e.g., for things like course-wide code syntax highlighting):
+
+    "Having the assets tied to the course means that even if the block [that uses these assets] is removed, the assets would still be loaded. It would be better to have them requested on-demand by only the blocks that need them. Having said that, we don't have a mechanism in XBlock to allow multiple blocks to share the same library."
+
+This is a valid point; there's a risk of vestigial libraries being loaded with this approach. The simplest way to address this concern is to recommend that authors not use this feature for libraries and instead continue loading these using the existing mechanism (i.e., by adding script tags in any blocks where they are needed).
+
+Another comment mentioned the impact of additional HTTP requests:
+
+    "There are performance implications to loading a number of individual files like this. Having said that, it would be difficult to have individual courses contribute files to the static asset pipeline, since courses can be created/imported after the LMS has been stood up."
+
+Some benchmarking will likely be appropriate here, though hopefully the benefits of this feature will outweigh any performance impact if only a few scripts are included. Caching should help to reduce the duration of each request after the initial load: it looks like edX sets the Cache-Control max-age to 1 year, and external JavaScript CDNs should also have reasonable caching behavior. Hopefully HTTP/2 will also help as it gains adoption.
```

Comment:
Note: only files with MD5 hashes are cached with a 1 year lifetime. Files that have no MD5 hash are uncached as there is no ability to know whether the content has changed. We would need to find a way to get these files to be served with an MD5 hash in order to benefit from caching. It is probably possible to model this on the way XBlocks serve their assets.